File: postinst.c
#include <spawn.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
extern char **environ;

int run_cmd(char *cmd)
{
    pid_t pid;
    char *argv[] = {"sh", "-c", cmd, NULL};
    int status = posix_spawn(&pid, "/bin/sh", NULL, NULL, argv, environ);
    if (status == 0) {
        if (waitpid(pid, &status, 0) == -1) {
            perror("waitpid");
        }
    }
    return status;
}

int main()
{
    if (geteuid() != 0) {
        printf("Run this as root!\n");
        exit(1);
    }
    
    run_cmd("launchctl unload /Library/LaunchDaemons/com.openssh.sshd.plist");
    if (access("/usr/libexec/sshd-keygen-wrapper", F_OK) == 0) {
        run_cmd("plutil -key Sockets -key LocalSSHListener -dict /Library/LaunchDaemons/com.openssh.sshd.plist");
        run_cmd("plutil -key Sockets -key LocalSSHListener -key SockServiceName -string 2222 /Library/LaunchDaemons/com.openssh.sshd.plist");
    } else {
        run_cmd("sed -i 's/#Port 22/Port 22/' /etc/ssh/sshd_config");
        run_cmd("sed -i '/Port 22/a\\Port 2222' /etc/ssh/sshd_config");
    }
    run_cmd("launchctl load /Library/LaunchDaemons/com.openssh.sshd.plist");
    return 0;
}

File: postinst.m
#import <Foundation/Foundation.h>

bool modifyPlist(NSString *filename, void (^function)(id)) {
    NSData *data = [NSData dataWithContentsOfFile:filename];
    if (data == nil) {
        return false;
    }
    NSPropertyListFormat format = 0;
    id plist = [NSPropertyListSerialization propertyListWithData:data options:NSPropertyListMutableContainersAndLeaves format:&format error:nil];
    if (plist == nil) {
        return false;
    }
    if (function) {
        function(plist);
    }
    NSData *newData = [NSPropertyListSerialization dataWithPropertyList:plist format:format options:0 error:nil];
    if (newData == nil) {
        return false;
    }
    if (![data isEqual:newData]) {
        if (![newData writeToFile:filename atomically:YES]) {
            return false;
        }
    }
    return true;
}

void run_system(const char *cmd) {
    int status = system(cmd);
    if (WEXITSTATUS(status) != 0) {
        printf("Error in command: \"%s\"\n", cmd);
        exit(WEXITSTATUS(status));
    }
}

int main() {
    if (getuid() != 0) {
        printf("Run this as root!\n");
        return 1;
    }

    run_system("launchctl unload /Library/LaunchDaemons/com.openssh.sshd.plist");
    NSMutableDictionary *LocalSSHListener = [NSMutableDictionary new];
    LocalSSHListener[@"SockNodeName"] = @"localhost";
    LocalSSHListener[@"SockServiceName"] = @"2222";
    NSString *const plist = @"/Library/LaunchDaemons/com.openssh.sshd.plist";
    modifyPlist(plist, ^(id plist) {
        plist[@"Sockets"][@"LocalSSHListener"] = LocalSSHListener;
    });
    run_system("launchctl load /Library/LaunchDaemons/com.openssh.sshd.plist");
    return 0;
}
